flask/wrappers.py

```diff
@@ -18,14 +18,93 @@
+        :mimetype:`application/json` or :mimetype:`application/*+json`.
+        """
+        mt = self.mimetype
+        if mt == 'application/json':
+            return True
+        if mt.startswith('application/') and mt.endswith('+json'):
+            return True
+        return False
+
+    @property
+    def json(self):
+        """If the request/response is JSON then this will contain the parsed
+        JSON data.  Otherwise the :meth:`on_json_loading_failed` method will
+        get invoked.
+
+        This is deprecated, the :meth:`get_json` method should be used instead.
+        """
+        from warnings import warn
+        warn(DeprecationWarning('json is deprecated.  '
+                                'Use get_json() instead.'), stacklevel=2)
+        return self.get_json()
+
+    def _get_data_for_json(req, cache):
+        getter = getattr(req, 'get_data', None)
+        if getter is not None:
+            return getter(cache=cache)
+        return req.data
+
+    def get_json(self, force=False, silent=False, cache=True):
+        """Parses the incoming JSON request data and returns it.  If parsing
+        fails the :meth:`on_json_loading_failed` will be invoked. By default
+        this function will only load the JSON data if :attr:`is_json` is true,
+        but this can be overridden by the `force` parameter.
+
+        :param force: if set to ``True`` the mimetype is ignored.
+        :param silent: if set to ``True`` this method will fail silently
+                       and return ``None``.
+        :param cache: if set to ``True`` the parsed JSON data is remembered
+                      on the request.
+        """
+        rv = getattr(self, '_cached_json', _missing)
+        if rv is not _missing:
+            return rv
+
+        if not (force or self.is_json):
+            return None
+
+        # We accept a request charset against the specification as certain
+        # clients have been using this in the past.  For responses, we assume
+        # that if the response charset was set explicitly then the data had
+        # been encoded correctly as well.
+        charset = self.mimetype_params.get('charset')
+        try:
+            data = self._get_data_for_json(cache)
+            if charset is not None:
+                rv = json.loads(data, encoding=charset)
+            else:
+                rv = json.loads(data)
+        except ValueError as e:
+            if silent:
+                rv = None
+            else:
+                rv = self.on_json_loading_failed(e)
+        if cache:
+            self._cached_json = rv
+        return rv
+
+    def on_json_loading_failed(self, e):
+        """Called if decoding of the JSON data failed.  The return value of
+        this method is used by :meth:`get_json` when an error occurred.  The
+        default implementation just raises a :class:`BadRequest` exception.
+        """
+        raise BadRequest()
```

Comment:
One could use a single method that first checks whether a context is available before trying to access current_app.